docs/design/2022-03-09-optimize-gc-for-internal-transaction.md

```diff
@@ -0,0 +1,207 @@
+        close(exited)
+    })
+}
+```
+
+
+
+the main code for `InitInnerTxnStartTsBox()` is as follower. It starts `storeInnerTxnStartTsLoop` goroutine to store an internal transaction startTS in `globalInnerTxnTsBox.innerTxnStartTsMap`, starts `deleteInnerTxnStartTsLoop` goroutine to delete an internal transaction startTS from `globalInnerTxnTsBox.innerTxnStartTsMap`. In particular, It starts `processUndeletedStartTsLoop` to delete the startTS for some rarely scenes.
+
+```go
+func InitInnerTxnStartTsBox() {
+	iTxnTsBox := &innerTxnStartTsBox{
+		chanToStoreStartTS:     make(chan uint64, chanBufferSize),
+		chanToDeleteStartTS:    make(chan uint64, chanBufferSize),
+		innerTxnStartTsMap:     make(map[uint64]uint64, chanBufferSize),
+		chanToProcUndelStartTS: make(chan uint64, undeletedStartTsBufferSize),
+		undeletedStartTsMap:    make(map[uint64]uint64, undeletedStartTsBufferSize),
+	}
+	globalInnerTxnTsBox.Store(iTxnTsBox)
+	iTxnTsBox.wg.Add(3)
+	iTxnTsBox.innerTxnTsBoxRun.Store(true)
+	go iTxnTsBox.storeInnerTxnStartTsLoop()
+	go iTxnTsBox.deleteInnerTxnStartTsLoop()
+	go iTxnTsBox.processUndeletedStartTsLoop()
+}
+```
+
+
+
+#### store and delete internal sessions
+
+TiDB gets an internal session from a session pool implemented in `(s *session) getInternalSession`, when the transaction finished,it puts the session to the pool. This design stores the internal session to session manger and deletes the internal session from session manager in the function  `(s *session) getInternalSession`. It calls the function `infosync.DeleteInternalSession()` to delete the internal session from session manager and calls the function `infosync.StoreInternalSession` to add the internal session to session manger.
+
+```go
+func (s *session) getInternalSession(execOption sqlexec.ExecOption) (*session, func(), error) {
+	tmp, err := s.sysSessionPool().Get()
+	se := tmp.(*session)
+	// Put the internal session to the map of SessionManager
+	infosync.StoreInternalSession(unsafe.Pointer(se))
+	return se, func() {
+		// Delete the internal session to the map of SessionManager
+		infosync.DeleteInternalSession(unsafe.Pointer(se))
+		s.sysSessionPool().Put(tmp)
+    }
+}
+```
+
+
+
+#### store and delete internal transactions
+
+The internal transactions here are executed by function `RunInNewTxn`. It stores the internal transaction startTS by function  `wrapStoreInterTxnTS()` and deletes the transaction startTS by function `wrapDeleteInterTxnTS()`.
+
+```go
+// RunInNewTxn will run the f in a new transaction environment.
+func RunInNewTxn(ctx context.Context, store Storage, retryable bool, f func(ctx context.Context, txn Transaction) error) error {
+	defer func() {
+		wrapDeleteInterTxnTS(originalTxnTS)
+	}()
+    for i := uint(0); i < maxRetryCnt; i++ {
+        txn, err = store.Begin()
+        		if i == 0 {
+			originalTxnTS = txn.StartTS()
+			wrapStoreInterTxnTS(originalTxnTS)
+		}
+		err = txn.Commit(ctx)
+        ..........................................
```

Comment:
Replace indent space with tab, and IMO, 3 dots are enough to represent omitted codes.

Reply:
Got it